File: web/src/api/v2/watchTogetherSuggestions.ts
import {
  captureProfileRequestContext,
  isCapturedProfileAuthorityActive,
  StaleApiRequestContextError,
  type ProfileRequestContextSnapshot,
} from "@/api/client";
import type { WatchTogetherSuggestion } from "@/lib/watchTogether";
import { v2 } from "./request";

function requireAuthority(
  context: ProfileRequestContextSnapshot | null,
): asserts context is ProfileRequestContextSnapshot {
  if (!context || !isCapturedProfileAuthorityActive(context))
    throw new StaleApiRequestContextError();
}
export async function listRoomSuggestions(
  roomId: string,
  roomToken: string,
  context = captureProfileRequestContext(),
) {
  requireAuthority(context);
  const items: WatchTogetherSuggestion[] = [];
  const ids = new Set<string>(),
    cursors = new Set<string>();
  let cursor: string | undefined;
  for (let page = 0; page < 100; page++) {
    requireAuthority(context);
    const result = await v2("GET /api/v2/watch-together/rooms/{room_id}/suggestions", {
      path: { room_id: roomId },
      headers: { "X-Room-Token": roomToken },
      query: { limit: 100, cursor },
      profileContext: context,
    });
    requireAuthority(context);
    for (const row of result.items) {
      const user = Number(row.suggester_user_id);
      if (
        !Number.isSafeInteger(user) ||
        user <= 0 ||
        ids.has(row.id) ||
        row.room_id !== roomId ||
        !["movie", "episode"].includes(row.content_type)
      )
        throw new Error("Invalid room suggestion.");
      ids.add(row.id);
      items.push({
        ...row,
        suggester_user_id: user,
        content_type: row.content_type as "movie" | "episode",
      });
    }
    if (!result.page) throw new Error("Missing suggestion continuation.");
    if (!result.page.has_more) {
      if (result.page.next_cursor) throw new Error("Invalid suggestion continuation.");
      items.sort(
        (a, b) =>
          b.vote_count - a.vote_count ||
          Date.parse(a.created_at) - Date.parse(b.created_at) ||
          a.id.localeCompare(b.id),
      );
      return { suggestions: items };
    }
    const next = result.page.next_cursor;
    if (!next || cursors.has(next)) throw new Error("Invalid suggestion continuation.");
    cursors.add(next);
    cursor = next;
  }
  throw new Error("Too many room suggestions. Reload the room.");
}
export async function setRoomSuggestionVote(
  roomId: string,
  roomToken: string,
  suggestionId: string,
  vote: boolean,
) {
  const context = captureProfileRequestContext();
  requireAuthority(context);
  const operation = vote
    ? "POST /api/v2/watch-together/rooms/{room_id}/suggestions/{suggestion_id}/vote"
    : "DELETE /api/v2/watch-together/rooms/{room_id}/suggestions/{suggestion_id}/vote";
  await v2(operation, {
    path: { room_id: roomId, suggestion_id: suggestionId },
    headers: { "X-Room-Token": roomToken },
    profileContext: context,
    retryAuthentication: false,
  });
  requireAuthority(context);
  return listRoomSuggestions(roomId, roomToken, context);
}

File: internal/apiv2/watch_together_suggestions.go
package apiv2

import (
	"context"
	"errors"
	"net/http"
	"strconv"

	"github.com/Silo-Server/silo-server/internal/watchtogether"
)

type WatchTogetherSuggestionService interface {
	CheckSuggestionRoomProof(string, int, string, string) error
	ListSuggestionPage(context.Context, string, string, int, *watchtogether.SuggestionPosition) ([]watchtogether.Suggestion, bool, error)
	SetSuggestionVote(context.Context, string, string, int, string, bool) error
}

type WatchTogetherSuggestionListInput struct {
	RoomID    string `path:"room_id"`
	RoomToken string `header:"X-Room-Token" maxLength:"4096"`
	LimitParam
	Cursor string `query:"cursor"`
}
type WatchTogetherSuggestionVoteInput struct {
	RoomID       string `path:"room_id"`
	SuggestionID string `path:"suggestion_id"`
	RoomToken    string `header:"X-Room-Token" maxLength:"4096"`
}
type WatchTogetherSuggestion struct {
	ID                 ID      `json:"id"`
	RoomID             ID      `json:"room_id"`
	SuggesterUserID    ID      `json:"suggester_user_id"`
	SuggesterProfileID ID      `json:"suggester_profile_id"`
	ContentID          ID      `json:"content_id"`
	ContentType        string  `json:"content_type" enum:"movie,episode"`
	Title              string  `json:"title"`
	Subtitle           string  `json:"subtitle"`
	PosterURL          string  `json:"poster_url"`
	Note               string  `json:"note"`
	VoteCount          int     `json:"vote_count"`
	VotedByMe          bool    `json:"voted_by_me"`
	CreatedAt          Instant `json:"created_at"`
}
type WatchTogetherSuggestionListOutput struct {
	Body Collection[WatchTogetherSuggestion]
}

func suggestionProblem(err error) error {
	switch {
	case errors.Is(err, watchtogether.ErrRoomNotFound), errors.Is(err, watchtogether.ErrSuggestionNotFound):
		return NewProblem(TypeNotFound, "The room or suggestion was not found.")
	case errors.Is(err, watchtogether.ErrRoomClosed):
		return NewProblem(TypeConflict, "The room is closed.")
	default:
		return serviceProblem(err)
	}
}
func (reg *Registry) suggestionProof(ctx context.Context, room, token string) (int, string, error) {
	if reg.deps.WatchTogetherSuggestions == nil {
		return 0, "", unavailable("watch together")
	}
	user, profile, p := viewerIdentity(ctx)
	if p != nil {
		return 0, "", p
	}
	if err := reg.deps.WatchTogetherSuggestions.CheckSuggestionRoomProof(room, user, profile, token); err != nil {
		return 0, "", suggestionProblem(err)
	}
	return user, profile, nil
}
func registerWatchTogetherSuggestions(reg *Registry) {
	const root = Prefix + "/watch-together/rooms/{room_id}/suggestions"
	const listID = "listWatchTogetherSuggestions"
	const creationOrder = "created_at,id"
	cursors := NewCursors(reg.deps.CursorSecret)
	list := Operation{Operation: humaOp(http.MethodGet, root, listID, "realtime", "List room suggestions in bounded creation order. Vote changes do not reorder the live traversal; this is not a snapshot."), Class: ClassProfileScoped, ServiceBacked: true}
	Register(reg, list, func(ctx context.Context, in *WatchTogetherSuggestionListInput) (*WatchTogetherSuggestionListOutput, error) {
		user, profile, err := reg.suggestionProof(ctx, in.RoomID, in.RoomToken)
		if err != nil {
			return nil, err
		}
		scope := CursorScope{OperationID: listID, Security: strconv.Itoa(user) + ":" + profile + ":" + viewerScopeDigest(ctx), Filter: in.RoomID + ":" + strconv.Itoa(in.Limit), Sort: creationOrder, Tiebreaker: "id"}
		var after *watchtogether.SuggestionPosition
		if in.Cursor != "" {
			after = new(watchtogether.SuggestionPosition)
			if p := cursors.Decode(scope, in.Cursor, after); p != nil {
				return nil, p
			}
		}
		rows, more, err := reg.deps.WatchTogetherSuggestions.ListSuggestionPage(ctx, in.RoomID, profile, in.Limit, after)
		if err != nil {
			return nil, suggestionProblem(err)
		}
		items := make([]WatchTogetherSuggestion, 0, len(rows))
		for _, s := range rows {
			items = append(items, WatchTogetherSuggestion{ID: ID(s.ID), RoomID: ID(s.RoomID), SuggesterUserID: ID(strconv.Itoa(s.SuggesterUserID)), SuggesterProfileID: ID(s.SuggesterProfileID), ContentID: ID(s.ContentID), ContentType: s.ContentType, Title: s.Title, Subtitle: s.Subtitle, PosterURL: s.PosterURL, Note: s.Note, VoteCount: s.VoteCount, VotedByMe: s.VotedByMe, CreatedAt: NewInstant(s.CreatedAt)})
		}
		next := ""
		if more {
			if len(rows) == 0 {
				return nil, unavailable("suggestion pagination")
			}
			last := rows[len(rows)-1]
			next, err = cursors.Encode(scope, watchtogether.SuggestionPosition{CreatedAt: last.CreatedAt, ID: last.ID})
			if err != nil {
				return nil, serviceProblem(err)
			}
		}
		return &WatchTogetherSuggestionListOutput{Body: Collection[WatchTogetherSuggestion]{Items: items, Page: &PageInfo{HasMore: more, NextCursor: next}}}, nil
	})
	for _, v := range []struct {
		method, id string
		vote       bool
	}{{http.MethodPost, "voteWatchTogetherSuggestion", true}, {http.MethodDelete, "unvoteWatchTogetherSuggestion", false}} {
		op := Operation{Operation: humaOp(v.method, root+"/{suggestion_id}/vote", v.id, "realtime", "Set this profile's vote membership. An already satisfied vote state returns an empty receipt; opposing writes are not ordered."), Class: ClassProfileScoped, ServiceBacked: true, DemoRestricted: isMutatingMethod(v.method), RetrySafety: RetrySafetyNaturalIdempotent}
		if v.vote {
			op.RetrySafety = RetrySafetyUniqueConstraint
		}
		op.DefaultStatus = http.StatusNoContent
		op.Errors = []int{409}
		Register(reg, op, func(ctx context.Context, in *WatchTogetherSuggestionVoteInput) (*struct{}, error) {
			user, profile, err := reg.suggestionProof(ctx, in.RoomID, in.RoomToken)
			if err != nil {
				return nil, err
			}
			if err = reg.deps.WatchTogetherSuggestions.SetSuggestionVote(ctx, in.RoomID, in.SuggestionID, user, profile, v.vote); err != nil {
				return nil, suggestionProblem(err)
			}
			return &struct{}{}, nil
		})
	}
}
